src/server/transaction.cc

```diff
@@ -1245,9 +1245,16 @@ bool Transaction::CancelShardCb(EngineShard* shard) {
     shard->shard_lock()->Release(LockMode());
   } else {
     auto lock_args = GetLockArgs(shard->shard_id());
-    DCHECK(sd.local_mask & KEYLOCK_ACQUIRED);
-    DCHECK(!lock_args.fps.empty());
-    GetDbSlice(shard->shard_id()).Release(LockMode(), lock_args);
+
+    if (lock_args.fps.empty()) {
+      // For NO_KEY_TRANSACTIONAL commands, we don't need to release locks
+      // because we don't acquire locks for them.
+      DCHECK(cid_->opt_mask() & CO::NO_KEY_TRANSACTIONAL);
+    } else {
+      DCHECK(sd.local_mask & KEYLOCK_ACQUIRED);
+      GetDbSlice(shard->shard_id()).Release(LockMode(), lock_args);
+    }
```

Comment:
Almost. I think the condition should be reversed:

```suggestion
    if (cid_->opt_mask() & CO::NO_KEY_TRANSACTIONAL) == 0) {
      auto lock_args = GetLockArgs(shard->shard_id());
      DCHECK(sd.local_mask & KEYLOCK_ACQUIRED);
      DCHECK(!lock_args.fps.empty());
      GetDbSlice(shard->shard_id()).Release(LockMode(), lock_args);
    }
```